contracts/traits/diamond/mod.rs

```diff
@@ -0,0 +1,86 @@
+pub use crate::traits::{
+    errors::{
+        DiamondError,
+        OwnableError,
+    },
+    ownable::*,
+};
+use brush::traits::Hash;
+use ink_prelude::vec::Vec;
+
+#[brush::wrapper]
+pub type DiamondRef = dyn Diamond;
+
+pub type Selector = [u8; 4];
+
+pub enum FacetCutAction {
+    Add = 0,
+    Replace = 1,
+    Remove = 2,
+    Unknown = 3,
+}
+
+impl From<u8> for FacetCutAction {
+    fn from(number: u8) -> Self {
+        return match number {
+            0 => FacetCutAction::Add,
+            1 => FacetCutAction::Replace,
+            2 => FacetCutAction::Remove,
+            _ => FacetCutAction::Unknown,
+        }
+    }
+}
+
+/// Struct which we use to initialize/update/remove a facet in the diamond
+#[derive(Default, Debug, Clone, PartialEq, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub struct FacetCut {
+    /// The `hash` of the code that should be executed.
+    pub hash: Hash,
+    /// The selector bytes that identify the function that should be called.
+    pub selectors: Vec<([u8; 4], u8)>,
+}
+
+/// Struct which we use to initialize the diamond contract
+#[derive(Default, Debug, Clone, PartialEq, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+pub struct InitCall {
+    /// The `hash` of the code that should be executed.
+    pub hash: Hash,
+    /// The selector bytes that identify the function that should be called.
+    pub selector: Selector,
+    /// The SCALE encoded parameters that are passed to the called function.
+    pub input: Vec<u8>,
+}
+
+/// Trait to be implemented in the contract which holds the diamond storage
+/// Also implements functions of Diamond Loupe to lookup the functionality of the diamond contract
+#[brush::trait_definition]
+pub trait Diamond {
```

Comment:
Diamond Loupe should be an extension for Diamond. So it should have separate storage=) If you need to override the internal function to track adding/replacing/removing selectors you can check the PR with PSP34 enumerable. It is resolved there.